pkg/capacity/manager.go

```diff
@@ -118,11 +133,16 @@ func (b *ClusterManager) GetOptimalTargetCluster(preferredCluster, namespace str
 	return optimalTargetClusters[0], nil
 }
 
-func getOptimalTargetClusters(preferredCluster string, getMemberClusters cluster.GetMemberClustersFunc, conditions ...cluster.Condition) []string {
+// getOptimalTargetClusters checks if a preferred target cluster was provided and available from the cluster pool.
+// If the preferred target cluster was not provided or not available, but a list of clusterRoles was provided, then it filters only the available clusters matching all those roles.
+// If no cluster roles were provided then it returns all the available clusters.
+// The function returns a slice with an empty string if not optimal target clusters where found.
+func getOptimalTargetClusters(preferredCluster string, getMemberClusters cluster.GetMemberClustersFunc, clusterRoles []string, conditions ...cluster.Condition) []string {
+	emptyTargetCluster := []string{""}
 	// Automatic cluster selection based on cluster readiness
 	members := getMemberClusters(append(conditions, cluster.Ready)...)
 	if len(members) == 0 {
-		return []string{""}
+		return emptyTargetCluster
 	}
 
 	// if the preferred cluster is provided and it is also one of the available clusters, then the same name is returned, otherwise, it returns the first available one
```

Comment:
personally, I think that the roles should have precedence over the preferred cluster. The preferred target cluster is the cluster where the user was provisioned last time. It's not the cluster that is specified in the Spec. Imagine:
1. you are provisioned to member1
2. you get deactivated
3. when you reactivate your account, then we try our best to provision you to member1 again. 
4. but, if the cluster is full or is gone then we provision you somewhere else

So if:
* either the purpose of the cluster has changed (we removed a label from the cluster)
* or the required set of labels has changed for your Space

Then we should not provision the Space to the same member cluster but somewhere else that matches the requirements

Reply:
Sorry I got confused between `targetCluster` and `preferredCluster` , changed the logic to first filter on cluster roles. see https://github.com/codeready-toolchain/host-operator/pull/736/commits/6b01926137a7f5c783ed0b886e61eaba8cc13642